index.bs

```diff
@@ -4919,7 +4922,7 @@ Authenticators may be required to store arbitrary strings chosen by a [=[RP]=],
 
 ### String Truncation ### {#sctn-strings-truncation}
 
-Each arbitrary string in the API will have some accommodation for the potentially limited resources available to an [=authenticator=]. If string value truncation is the chosen accommodation then authenticators MAY truncate in order to make the string fit within a length equal or greater than the specified minimum supported length. Such truncation SHOULD also respect UTF-8 sequence boundaries or [=grapheme cluster=] boundaries [[UAX29]]. This defines the maximum truncation permitted and authenticators MUST NOT truncate further.
+Each arbitrary string in the API will have some accommodation for the potentially limited resources available to an [=authenticator=]. If string value truncation is the chosen accommodation then authenticators MAY truncate in order to make the string fit within a length equal to or greater than the specified minimum supported length. Such truncation SHOULD also respect UTF-8 sequence boundaries or [=grapheme cluster=] boundaries [[UAX29]]. This defines the maximum truncation permitted and authenticators MUST NOT truncate further.
```

Comment:
I think "UTF-8 sequence boundaries" is unclear. I would suggest saying "UTF-8 character boundaries" or "UTF-8 code point boundaries". I would also make SHOULD->MUST. Arbitrary truncation of UTF-8 mid-character should not be permitted. I'm okay with SHOULD for grapheme cluster boundaries. Ultimately, I'd make that sentence say:

> Such truncation MUST only be done on UTF-8 character boundaries 
> and SHOULD respect [=grapheme cluster]= boundaries [[UAX29]].

Note that the "equal or greater to" discussion is super tricky. Because UTF-8 is a multibyte encoding, the truncation to (say) 64-bytes might end up with a string that is 62 or 63 bytes long (so as to avoid mid-character truncation).

Reply:
> I think "UTF-8 sequence boundaries" is unclear. I would suggest saying "UTF-8 character boundaries" or "UTF-8 code point boundaries".

@agl thoughts on this?

> I would also make SHOULD->MUST.

I believe we wrote this as a SHOULD instead of a MUST because there already exist authenticator implementations that do not respect these guidelines. This is also why we recommend that the client performs this truncation instead, because clients are better equipped to handle all the Unicode intricacies like identifying grapheme clusters that span multiple code points.

> Note that the "equal or greater to" discussion is super tricky. Because UTF-8 is a multibyte encoding, the truncation to (say) 64-bytes might end up with a string that is 62 or 63 bytes long (so as to avoid mid-character truncation).

Note that what's specified here with "equal to or greater than" is the _length the string may be truncated to fit within_: authenticators may truncate strings to _fit within_ X bytes, where X is "equal to or greater than the specified minimum supported length", so in this case X >= 64. So what this says is that authenticators may truncate the string _to fit within 64 bytes_, not just to exactly 64 bytes, for the reasons you point out. But it also gives authenticators the option to choose X _higher_ than 64 if they like. Then there's also the next requirement: "This defines the maximum truncation permitted and authenticators MUST NOT truncate further.", so an (X+1)-byte string may be truncated to X-3 bytes but not to, say, X-32 bytes.

Reply:
> Note that what's specified here with "equal to or greater than" is the length the string may be truncated to fit within: authenticators may truncate strings to fit within X bytes, where X is "equal to or greater than the specified minimum supported length", so in this case X >= 64. So what this says is that authenticators may truncate the string to fit within 64 bytes, not just to exactly 64 bytes, for the reasons you point out. But it also gives authenticators the option to choose X higher than 64 if they like. Then there's also the next requirement: "This defines the maximum truncation permitted and authenticators MUST NOT truncate further.", so an (X+1)-byte string may be truncated to X-3 bytes but not to, say, X-32 bytes.

I have to admit that the "equal to or greater than" phrasing is hard to parse. I read this (as intended) as allowing an implementation to choose a longer limit but not one that is shorter than the given limit (64 in most of our examples).

It's possible that people will read the "MUST NOT truncate further" requirement as meaning "exactly X bytes", when UTF-8 will produce some truncated strings of less than X bytes (but never more than 3 fewer).

> I believe we wrote this as a SHOULD instead of a MUST because there already exist authenticator implementations that do not respect these guidelines.

Perhaps there are such authenticators: they should feel icky and awful and fix their code 😉. I don't see why your spec should not require it. Note that I do not think we should hold up the same bar for grapheme cluster boundaries.

Reply:
> Would this phrasing be better?

>> Authenticators MUST NOT truncate further than the minimum amount required to align with a [=grapheme cluster=] boundary.

That's pretty good.

If we took this I would still make the change I requested to say "UTF-8 character boundaries" or "UTF-8 code point boundaries" in the preceding sentence.

---

Regarding UTF-8 truncation, I understand what you're saying and my comment was (perhaps inappropriately) dismissive of legacy implementations. 

"UTF-8 engine" makes it sound complicated. The design of UTF-8 is such that truncation on character boundaries can rely on a bitmask to check the byte at the truncation location. The code doesn't have to have a deep (or even shallow) understanding of Unicode to find code point boundaries. 

My concern about using SHOULD here is that it leaves the wrong impression. It might allow _more_ of these bad actors to come into the world. If we must preserve compatibility with these stale implementations (and I agree that we probably do), we should look for a formulation that does what I suggested before: makes people feel icky about not fixing their code.

Perhaps we can sidestep the problem:

```suggestion
Each arbitrary string in the API will have some accommodation for the potentially limited resources available to an
[=authenticator=]. If string value truncation is the chosen accommodation then authenticators MAY truncate the 
string value using a size limit equal to or greater than the specified minimum supported length. The resulting 
truncated value MAY be shorter than the chosen limit only by the number of bytes necessary to ensure that the
truncation falls on [=code point=] or [=grapheme cluster=][[UAX29]] boundaries in the UTF-8 encoded text. 
```

Reply:
> "UTF-8 engine" makes it sound complicated. The design of UTF-8 is such that truncation on character boundaries can rely on a bitmask to check the byte at the truncation location. The code doesn't have to have a deep (or even shallow) understanding of Unicode to find code point boundaries.

This is true of code points, but the greater point of this section is that Unicode code point boundaries don't necessarily align with what a user considers a "character". See section [3 Grapheme Cluster Boundaries](https://www.unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries) in our referenced [UAX#29]:

>[...] For example, “G” + _grave-accent_ is a _user-perceived character_: users think of it as a single character, yet is actually represented by two Unicode code points. [...]

See also the reference to [grapheme cluster](https://w3c.github.io/i18n-glossary/#dfn-grapheme-cluster) in the Internationalization Glossary:
>[...] For example, a Slovak user might wish to treat the default pair of grapheme clusters "ch" as a single grapheme cluster. Note that the interaction between the language of string content and the end-user's preferences might be complex.

This is what I meant vaguely by "UTF-8 engine" earlier: the default rules for grapheme cluster segmentation are already complex, but a truly faithful implementation should also account for the user's preferences, locale settings and possibly even other context information. An authenticator separate from the client (user agent) cannot feasibly take all of this into account. The client is much more likely to both have enough storage and update capabilities to keep up with these language processing rules and to have access to the necessary context information in the first place.

All this in mind, I came up with this formulation. I feel like this may be a bit more stuffy than necessary, but at least it should be unambiguous. What do you think?

>If string truncation is the chosen accommodation, then the authenticator MAY choose a size limit equal to or greater than the specified minimum supported length and truncate the string to fit within that limit. Such truncation SHOULD also respect UTF-8 code point boundaries or [grapheme cluster](https://w3c.github.io/i18n-glossary/#dfn-grapheme-cluster) boundaries [[UAX29]](file:///home/emlun/dev/webauthn/index.html#biblio-uax29). The resulting truncated value MAY be shorter than the chosen size limit but MUST NOT be shorter than the longest prefix substring that satisfies the size limit and ends on a [grapheme cluster](https://w3c.github.io/i18n-glossary/#dfn-grapheme-cluster) boundary.

Reply:
Thanks for your comments! I've rewritten the section to more clearly distinguish between the responsibilities of clients vs. authenticators, where the requirements are now MUST/SHOULD (code points/clusters) for clients and SHOULD/MAY for authenticators. Please have another look! I'll resolve this thread since the whole text is rearranged.